### ISSUE TITLE
Access ValueType parameters and ValueType __result if the type is unknown at compile time

### ISSUE THREAD

**Issue by Meigyoku-Thmn** [pardeike/Harmony#195]
**Is your feature request related to a problem? Please describe.**
In many cases, we cannot static link our app with the target program that has the method we want to patch, so we cannot know some struct type at compile time.

Suppose I have a method I want to patch:
```cs
public struct Unk {
   public string Str;
}
public class Target {
   public Unk CloneUnk(ref Unk unk) {
       var rs = new Unk();
       rs.Str = unk.Str;
       return rs;
   }
}
```
The following patch will not work:
```cs
static Type UnkType = GetUnkType();
static void CloneUnk(ref dynamic unk, ref dynamic __result) {
    dynamic rs = Activator.CreateInstance(UnkType, new object[]{});
    rs.Str = "Patched_" + unk.Str;
    __result = rs;
}
```
(If you set a breakpoint, you will see that dynamic unk and dynamic __result are shown that is cannot be read, and the program will just crash)

**Describe the solution you'd like**
I don't know how the solution would be.

**Describe alternatives you've considered**
Just use transpiler and hardcode some instructions (too cumbersome and hard-to-maintain).

**Additional context**
Add any other context or related material about the feature request here.

**Comment by pardeike** [comment]
Hi. I will investigate if there is a viable solution to this. In the meantime, I can tell you what I think happens when you use the “dynamic” keyword: the compiler generates as many copies of the method as necessary to match any usage of it for a given struct type. 

This is an assumption and I need to find out more about the dynamic keyword. If it’s true, then there is no easy way to support this in Harmony because when you compile your patch the usage is of course unknown to the compiler. With classes you can use “object” but I am not sure how to handle this with structs. 

I’ll get back to you on this when I have more time to investigate.

**Comment by Meigyoku-Thmn** [comment]
Ok thanks :D , so for now I will use transpiler as the alternative solution.

**Comment by pardeike** [comment]
Here is a basic example on how the compiler deals with dynamic:
https://gist.github.com/pardeike/fbb340c312dd29035c07064c474af30c

An interesting observation is that at the place where M3(f); is called, the compiler boxes the argument (which is just of object type) to make it an object. Now, if the original application has methods that use ref to pass structs then you cannot simply use “object” or “dynamic” in your patch and Harmony cannot either. Thus Harmony will generate the dynamic method to match the runtime information (the correct struct type) and you won’t get a problem in replacing the original method with the newly constructed method. The problem instead is that Harmony assume that you know the type you are dealing with and will insert calls to your patches that pass the arguments and instance as their original types. 

The question now is how Harmony should know when to box any structs. That would depend on if it is possible to detect your usage of “dynamic” in the patch arguments. I guess not, so that leaves only one options: Harmony would need to detect if you use “object” as argument type and box any structs. The problem with this is that it would fail if users use “object” as a type by error or by directly specifying it instead of using “dynamic” which would not generate all the extra code you see in my gist. 

It’s a bit tricky to decide. I could make Harmony simply box any valuetype if the destination type is object but that might lead to confusion for users that are not aware of how all this works.

**Comment by pardeike** [comment]
One more thing, it seems that ”ref dynamic” isn’t supported as you can see in this example:

https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUwMwAIDOwBOArgMbBYBiA9pVgN6paNaZZgB2ZbA3A0yzABYsAIQCG+ABQBKOlgC+qBWhT84WACoBTPHV6MWVGgFkE03SiaWsAN3FYAZlgC8WNpoDuFatJ4WrjewA6NmcsBDgMX399AHYHKKYlS34hIzgzej9/ABMATzZRAFswYgc1FxMfPSt7OECxSSkE/1r68XSm6qS+bEEsIwwJPILi0sKZTOisQrbG5qsZho757v1e1IEJfE1HYaKS6Ynqy0XxKqyF2eWu6pT+gFYM46ZDB1C3T0NzqYGJe06LpYjJtto5/itFEA=

**Comment by pardeike** [comment]
And for reference to myself:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/using-type-dynamic

**Comment by Meigyoku-Thmn** [comment]
I just wrote ref dynamic in hope that Dotnet would pass the struct argument as reference instead of copying it. But dynamic is not so important, "object" is good enough for me.

**Comment by pardeike** [comment]
I think **return type of struct of 8 bytes size or more** is fixed in Harmony 2.0b the remainder of this Issue will for now get a WONTFIX
